common/models/role.js

```diff
@@ -29,44 +33,66 @@ module.exports = function(Role) {
   // Set up the connection to users/applications/roles once the model
   Role.once('dataSourceAttached', function() {
     var roleMappingModel = this.RoleMapping || loopback.getModelByType(RoleMapping);
-    Role.prototype.users = function(callback) {
-      roleMappingModel.find({where: {roleId: this.id,
-        principalType: RoleMapping.USER}}, function(err, mappings) {
-        if (err) {
-          if (callback) callback(err);
-          return;
-        }
-        return mappings.map(function(m) {
-          return m.principalId;
-        });
-      });
-    };
+    var principalTypesToModels = {};
+
+    principalTypesToModels[RoleMapping.USER] = User;
+    principalTypesToModels[RoleMapping.APPLICATION] = Application;
+    principalTypesToModels[RoleMapping.ROLE] = Role;
+
+    Object.keys(principalTypesToModels).forEach(function(principalType){
+      var model = principalTypesToModels[principalType];
+      var pluralName = model.pluralModelName.toLowerCase();
+      /** 
+       * Fetch all users assigned to this role
+       * @function Role.prototype#users
+       * @param {object} [query] query object passed to model find call
+       * @param  {Function} [callback]
+       */ 
+      /** 
+       * Fetch all applications assigned to this role
+       * @function Role.prototype#applications
+       * @param {object} [query] query object passed to model find call
+       * @param  {Function} [callback]
+       */
+      /** 
+       * Fetch all roles assigned to this role
+       * @function Role.prototype#roles
+       * @param {object} [query] query object passed to model find call
+       * @param {Function} [callback]
+       */ 
+      Role.prototype[pluralName] = function(query, callback) {
+        listByPrincipalType(model, principalType, query, callback);
+      }
```

Comment:
consolidate most of the logic from users, applications, and roles into one function